modules/3d/include/opencv2/3d/depth.hpp

```diff
@@ -0,0 +1,1264 @@
+      RGBD_NORMALS_METHOD_FALS = 0,
+      RGBD_NORMALS_METHOD_LINEMOD = 1,
+      RGBD_NORMALS_METHOD_SRI = 2
+    };
+
+    RgbdNormals()
+        :
+          rows_(0),
+          cols_(0),
+          depth_(0),
+          K_(Mat()),
+          window_size_(0),
+          method_(RGBD_NORMALS_METHOD_FALS),
+          rgbd_normals_impl_(0)
+    {
+    }
+
+    /** Constructor
+     * @param rows the number of rows of the depth image normals will be computed on
+     * @param cols the number of cols of the depth image normals will be computed on
+     * @param depth the depth of the normals (only CV_32F or CV_64F)
+     * @param K the calibration matrix to use
+     * @param window_size the window size to compute the normals: can only be 1,3,5 or 7
+     * @param method one of the methods to use: RGBD_NORMALS_METHOD_SRI, RGBD_NORMALS_METHOD_FALS
+     */
+    RgbdNormals(int rows, int cols, int depth, InputArray K, int window_size = 5, int method =
+                RgbdNormals::RGBD_NORMALS_METHOD_FALS);
+
+    ~RgbdNormals();
+
+    CV_WRAP static Ptr<RgbdNormals> create(int rows, int cols, int depth, InputArray K, int window_size = 5, int method =
+                                           RgbdNormals::RGBD_NORMALS_METHOD_FALS);
+
+    /** Given a set of 3d points in a depth image, compute the normals at each point.
+     * @param points a rows x cols x 3 matrix of CV_32F/CV64F or a rows x cols x 1 CV_U16S
+     * @param normals a rows x cols x 3 matrix
+     */
+    CV_WRAP
+    void apply(InputArray points, OutputArray normals) const;
+
+    /** Initializes some data that is cached for later computation
+     * If that function is not called, it will be called the first time normals are computed
+     */
+    CV_WRAP
+    void initialize() const;
+
+    CV_WRAP
+    int getRows() const
+    {
+        return rows_;
+    }
+    CV_WRAP
+    void setRows(int val)
+    {
+        rows_ = val;
+    }
+    CV_WRAP
+    int getCols() const
+    {
+        return cols_;
+    }
+    CV_WRAP
+    void setCols(int val)
+    {
+        cols_ = val;
+    }
+    CV_WRAP
+    int getWindowSize() const
+    {
+        return window_size_;
+    }
+    CV_WRAP
+    void setWindowSize(int val)
+    {
+        window_size_ = val;
+    }
+    CV_WRAP
+    int getDepth() const
+    {
+        return depth_;
+    }
+    CV_WRAP
+    void setDepth(int val)
+    {
+        depth_ = val;
+    }
+    CV_WRAP
+    cv::Mat getK() const
+    {
+        return K_;
+    }
+    CV_WRAP
+    void setK(const cv::Mat &val)
+    {
+        K_ = val;
+    }
+    CV_WRAP
+    int getMethod() const
+    {
+        return method_;
+    }
+    CV_WRAP
+    void setMethod(int val)
+    {
+        method_ = val;
+    }
+
+  protected:
+    void
+    initialize_normals_impl(int rows, int cols, int depth, const Mat & K, int window_size, int method) const;
+
+    int rows_, cols_, depth_;
+    Mat K_;
+    int window_size_;
+    int method_;
+    mutable void* rgbd_normals_impl_;
```

Comment:
no "protected" stuff is allowed in "public headers".
Use `PImpl` instead.

Reply:
Finally fixed
(A lot of code in `depth.hpp` and related cpps had to be refactored)